src/python/ib1/openenergy/support/__init__.py

```diff
@@ -43,6 +46,236 @@ def httpclient_log(*args):
     http.client.HTTPConnection.debuglevel = 1
 
 
+def code_verifier(length: int = 128):
+    """
+    Return a cryptographically random string as specified in RFC7636 section 4.1
+
+    See https://datatracker.ietf.org/doc/html/rfc7636#section-4.1
+
+    :param length:
+        length of the generated string, minimum 43, maximum 128. Defaults to 128
+    :return:
+    """
+    vocab = string.ascii_letters + '-._~0123456789'
+    return ''.join([secrets.choice(vocab) for _ in range(length)])
+
+
+class CodeAuthMethod:
+    """
+    Additional auth logic passed to the FAPISession to enable code grant handling. Handles PKCE (RFC7636)
+    """
+
+    def __init__(self, redirect_uri: str, client_id: str):
+        """
+        Create a new code auth method with empty state, this needs to have a code created before use
+
+        :param redirect_uri:
+            The redirect back to the login callback
+        """
+        self.redirect_uri = redirect_uri
+        self.client_id = client_id
+        self._code = None
+        self._refresh_token = None
+        # RFC7636 logic (OAuth PKCE)
+        self._verifier = code_verifier()
+        self._code_challenge = CodeAuthMethod.code_challenge(self._verifier)
+        # State, re-use the verifier logic here because why not
+        self._state = code_verifier()
+
+    def as_jwt(self, secret: str) -> str:
+        """
+        Serialise the state of this code auth method to a JWT, encoding with the supplied symmetric key
+
+        :param secret:
+            Symmetric key used to encode the JWT
+        :returns:
+            Serialised JWT string represnting this object
+        """
+        d = {
+            'redirect_uri': self.redirect_uri,
+            'client_id': self.client_id,
+            'state': self._state,
+            'verifier': self._verifier,
+            'code_challenge': self._code_challenge
+        }
+        if self._code:
+            d['code'] = self._code
+        if self._refresh_token:
+            d['refresh_token'] = self._refresh_token
+        return jwt.encode(d, secret, algorithm='HS256')
```

Comment:
do we need any input validation here?

Reply:
It's not getting user input here, so I'd say no. The code and refresh tokens are values returned from the authz server within a JWT. Sure, someone could abuse this API and deliberately post their own values in here but it wouldn't do much. They're just opaque values, and are immediately sent to the authz server in exchange for an access token (in the case of the code). As far as I can tell the authz server we're using doesn't issue refresh tokens anyway. I might add a simple check that the lengths aren't insane to prevent some kind of potential attempt to crash things with excessive occupancy.